File: 정종길/프로그래머스/카카오 2회독/2024 KAKAO WINTER INTERNSHIP/(다시 풀기) n + 1 카드게임.py
def solution(coin, cards):
    
    answer = 0
    n = len(cards)
    idx = n // 3
    
    my_cards = set(cards[ : idx])
    get_cards = set()
    
    flag = True
    
    while flag:
        flag = False
        answer += 1
        
        for _ in range(2):
            if idx == n:
                return answer
            get_cards.add(cards[idx])
            idx += 1
            
        for card in my_cards:
            target = n + 1 - card
            if target in my_cards:
                my_cards.remove(card)
                my_cards.remove(target)
                flag = True
                break
            
            if coin >= 1 and target in get_cards:
                coin -= 1
                my_cards.remove(card)
                get_cards.remove(target)
                flag = True
                break
        
        if not flag and coin >= 2:
            for card in get_cards:
                target = n + 1 - card
                if target in get_cards:
                    coin -= 2
                    get_cards.remove(card)
                    get_cards.remove(target)
                    flag = True
                    break
    return answer

File: 정종길/프로그래머스/카카오 3회독/2024 KAKAO WINTER INTERNSHIP/(다시 풀기) n+1 카드게임.java
import java.util.*;

class Solution {
    int ccoin;
    
    public int solution(int coin, int[] cards) {
        int answer = 0;
        ccoin = coin;
        
        int n = cards.length;
        
        HashSet<Integer> havings = new HashSet<>();
        HashSet<Integer> candidates = new HashSet<>();
        
        for(int i = 0 ; i < n / 3; i++){
            havings.add(cards[i]);
        }

        int j = n / 3;
        
        while(j < n){
            answer += 1;
            for(int k = 0 ; k < 2; k++){
                candidates.add(cards[j]);
                j += 1;
            }
            
            if(!check_sum(havings, candidates, n)){
                return answer;
            }
        }
        
        answer += 1;
        
        return answer;
    }
    
    private Boolean check_sum(HashSet<Integer> havings, HashSet<Integer> candidates, int n){

        for(Integer h : havings){
            if(havings.contains(n + 1 - h)){
                havings.remove(h);
                havings.remove(n + 1 - h);
                return true;
            }
        }
        
        if(ccoin < 1){
            return false;
        }
        
        for(Integer h : havings){
            if(candidates.contains(n + 1 - h)){
                havings.remove(h);
                candidates.remove(n + 1 - h);
                ccoin -= 1;
                return true;
            }
        }
        
        if(ccoin < 2){
            return false;
        }
        
        for(Integer c : candidates){
            if(candidates.contains(n + 1 - c)){
                candidates.remove(c);
                candidates.remove(n + 1 - c);
                ccoin -= 2;
                return true;
            }
        }
        
        return false;
    }
}